### ISSUE TITLE
How to Read Consecutive Frames Extracted from Video Clips?

### ISSUE THREAD

**Issue by hitachinsk** [NVIDIA/DALI#1445]
Hello everyone, I'm a new comer to this project and want to modify the dataloader of a video-SR project with DALI because of the limitation of data-preprocessing with CPU.

Now, I have extracted my videos into frames and composed 5 frames as a group. Here is my file structure.
```
input
----video1
--------01
------------00001.jpg
------------00002.jpg
------------00003.jpg
------------00004.jpg
------------00005.jpg
--------02
------------00001.jpg
------------00002.jpg
------------00003.jpg
------------00004.jpg
------------00005.jpg
......
target
----video1
--------01
------------target1.jpg
--------02
------------target2.jpg
......
```
What I want is to read images from a video clip(for example `input/video1/01/*`) and corresponding target(for example `target/video1/target1.jpg`) in a pipeline and do data-processing to all of them.

Now I have to initialize 5 different input files and one ground-truth in my code in my code as 6 **ExternalSource**, and use open-cv to feed image data to them,  it seems something just like the following codes.
```
class HyperTrainPipe(Pipeline):
    def __init__(self, data_iterator, batch_size, num_threads, device_id, crop, local_rank=0, world_size=1, dali_cpu=False):
        super(Ai4kTrainPipe, self).__init__(batch_size, num_threads, device_id, seed=12)  # seed need to be modified
        self.iterator = data_iterator
        dali_device = 'gpu'
        self.input1 = ops.ExternalSource()
        self.input2 = ops.ExternalSource()
        self.input3 = ops.ExternalSource()
        self.input4 = ops.ExternalSource()
        self.input5 = ops.ExternalSource()
        self.gt = ops.ExternalSource()
```

I know this solution is too foolish, the [sequence pipeline](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/sequence_reader_simple_example.html?highlight=sequences) seems to be a solution to this problem, but how can I read the corresponding ground-truth with this pipeline? Or do you have other advice? I'm looking forward to your reply, THANK YOU!

**Comment by jantonguirao** [comment]
Hi @hitachinsk ,

You can use SequenceReader to read the input and FileReader/ImageDecoder to read the target images.

Here is a code example:
```
#!/bin/python

import os.path
import numpy as np
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types

input_path = os.path.join(os.environ['PWD'], 'test/input/video1')
target_path = os.path.join(os.environ['PWD'], 'test/target/video1')
batch_size = 2
sequence_length = 5
n_iter=10

class SequencePipeline(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, data_dir, target_dir, sequence_length,
                 reader_seed=123456):
        super(SequencePipeline, self).__init__(batch_size, num_threads, device_id, seed = 16)
        # Using same seed on both readers to produce the same order
        self.sequences = ops.SequenceReader(file_root = data_dir, sequence_length = sequence_length,
                                            seed=reader_seed)
        self.gt = ops.FileReader(file_root = target_dir, seed=reader_seed)
        self.decode = ops.ImageDecoder(device='cpu', output_type=types.RGB)

    def define_graph(self):
        seq = self.sequences()
        encoded, _ = self.gt()
        target = self.decode(encoded)
        return seq, target

pipe = SequencePipeline(batch_size = batch_size, num_threads = 2, device_id = 0,
                        data_dir = input_path, target_dir = target_path,
                        sequence_length = sequence_length)
pipe.build()


for i in range(n_iter):
    print("Iteration " + str(i))
    pipe_out = pipe.run()
    sequences_out = pipe_out[0].as_array()
    target_out = pipe_out[1].as_array()
    print(sequences_out.shape)
    print(target_out.shape)
```

I tested it with this directory structure (I just copied the same image for test purposes):
```
test
├── input
│   └── video1
│       ├── 01
│       │   ├── 00001.jpg
│       │   ├── 00002.jpg
│       │   ├── 00003.jpg
│       │   ├── 00004.jpg
│       │   └── 00005.jpg
│       └── 02
│           ├── 00001.jpg
│           ├── 00002.jpg
│           ├── 00003.jpg
│           ├── 00004.jpg
│           └── 00005.jpg
└── target
    └── video1
        ├── 01
        │   └── target1.jpg
        └── 02
            └── target2.jpg
```

Output:
```
read 2 files from 2 directories
Iteration 0
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 1
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 2
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 3
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 4
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 5
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 6
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 7
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 8
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
Iteration 9
(2, 5, 853, 1280, 3)
(2, 853, 1280, 3)
```

I hope it helps

**Comment by hitachinsk** [comment]
@jantonguirao Thanks for your reply! It seems that you set the seed of `self.sequence` and `self.gt` the same, is this the keypoint of reading video sequence and its ground-truth synchronously?

**Comment by JanuszL** [comment]
Hi,
Yes, it is about the seed. What is also important is that the number of possible sequences for the sequence reader is equal to the number of targets. Diving deeper:
- sequence reader reads all files, groups them by folder, treating as frames of video (one folder-> one video). Then based on `step`, `stride` and `sequence_length` arguments it generates possible sequences, then one sequence is treated as one sample. Then samples are shuffled in a predefined way (when you use shuffling) so you won't get samples from class 0, 1, ... n in order. Then they are read sequentially and when they are assembled into batch they are randomly sampled using this seed
- file reader - it is almost the same as above, but DALI just treats one file as one sample. So instead of sequences, you have target images in your case.
So as long directory names match between training and target, and seeds are the same as well then a sampling of sequences and targets will be done in the same order.

**Comment by hitachinsk** [comment]
Thanks for your kind help, now things work well for me.

**Comment by priyamdey** [comment]
Hi. One quick question: SequenceReader only support cpu as backend, not gpu. So, is the loading time same if I use a SequenceReader vs just looping over the images to read them as a sequence, using PIL?

**Comment by JanuszL** [comment]
Hi @priyamdey,

>  is the loading time same if I use a SequenceReader vs just looping over to read the images using PIL to form a sequence?

I think the reading process itself could be similar, but the overall data processing performance may very much depend on the operations you have latter in the pipeline. If you move that to the GPU and do resize/blur or any other operation then DALI may provide a performance benefit. Also, you can easily use the same pipeline in TensorFlow where it is hard to use PIL.
